pkg/checks/checks.go

```diff
@@ -0,0 +1,249 @@
+			continue
+		}
+		if err != nil {
+			log.Printf("the err was %#v\n", err)
+			return err
+		}
+		if resp.StatusCode == http.StatusOK {
+			return nil
+		}
+
+		select {
+		case <-ctx.Done():
+			log.Printf("ctx.Done called.  Returning err")
+			return ctx.Err()
+		case <-time.After(time.Second):
+		}
+	}
+}
+
+func checkNamespace(namespace string) bool {
+	whitelistedNamespaces := []string{
+		"openshift-",
+		"kube-",
+		"default",
+	}
+	for _, ns := range whitelistedNamespaces {
+		if strings.HasPrefix(namespace, ns) {
+			return true
+		}
+	}
+	return false
+}
+
+// WaitForInfraServices verify daemonsets, statefulsets
+func WaitForInfraServices(ctx context.Context, appclient *appclient.AppsV1Client) error {
+
+	// Daemonsets
+	dsList, err := appclient.DaemonSets("").List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, ds := range dsList.Items {
+		if !checkNamespace(ds.Namespace) {
+			continue
+		}
+		for {
+			ds, err := appclient.DaemonSets(ds.Namespace).Get(ds.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			if ds.Status.NumberMisscheduled > 0 {
+				return fmt.Errorf("Daemonset[%v] in Namespace[%v] has missscheduled[%v]", ds.Name, ds.Namespace, ds.Status.NumberMisscheduled)
+			}
+
+			if ds.Status.DesiredNumberScheduled == ds.Status.CurrentNumberScheduled && ds.Status.DesiredNumberScheduled == ds.Status.NumberReady {
+				break
+			}
+			select {
+			case <-time.After(2 * time.Second):
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+	}
+
+	// Statefulsets
+	ssList, err := appclient.StatefulSets("").List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, ss := range ssList.Items {
+		if !checkNamespace(ss.Namespace) {
+			continue
+		}
+		for {
+			ss, err := appclient.StatefulSets(ss.Namespace).Get(ss.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			if ss.Status.Replicas == ss.Status.ReadyReplicas && ss.Status.ReadyReplicas == ss.Status.CurrentReplicas {
+				break
+			}
+			select {
+			case <-time.After(2 * time.Second):
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+	}
+
+	// Deployments
+	dList, err := appclient.Deployments("").List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, d := range dList.Items {
+		if !checkNamespace(d.Namespace) {
+			continue
+		}
+		for {
+			d, err := appclient.Deployments(d.Namespace).Get(d.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			if d.Status.UnavailableReplicas > 0 {
+				return fmt.Errorf("Deployment[%v] in namespace[%v] has unavailablereplicas[%v]", d.Name, d.Namespace, d.Status.UnavailableReplicas)
+			}
+			if d.Status.Replicas == d.Status.ReadyReplicas && d.Status.ReadyReplicas == d.Status.AvailableReplicas {
+				break
+			}
+			select {
+			case <-time.After(2 * time.Second):
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+	}
+
+	// ReplicaSets
+	rsList, err := appclient.ReplicaSets("").List(metav1.ListOptions{})
```

Comment:
You shouldn't need to check for replicasets since you check for deployments, unless some replicasets run standalone.

Reply:
Currently they do not run standalone.  Is there any harm in doing so?

Reply:
My initial thought was that it's redundant since the deployment status api is an aggregation of all the replicasets under its control (I wrote that part :)). Now, I think it's also wrong because you may list old replicasets that should not have available replicas running.

Reply:
Will remove replicaset testing.